src/zw.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+#include "fault_injection.h"
+#include "usersim/zw.h"
+#include "utilities.h"
+#include <string>
+
+_IRQL_requires_max_(PASSIVE_LEVEL) NTSTATUS ZwCreateKey(
+    _Out_ PHANDLE key_handle,
+    _In_ ACCESS_MASK desired_access,
+    _In_ POBJECT_ATTRIBUTES object_attributes,
+    _Reserved_ ULONG title_index,
+    _In_opt_ PUNICODE_STRING class_string,
+    _In_ ULONG create_options,
+    _Out_opt_ PULONG disposition)
+{
+    if (usersim_fault_injection_inject_fault()) {
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    if (class_string != nullptr || title_index != 0) {
+        return STATUS_NOT_SUPPORTED;
+    }
+
+    UNREFERENCED_PARAMETER(create_options);
+
+    HKEY root_key = (HKEY)object_attributes->RootDirectory;
+    std::wstring relative_path = object_attributes->ObjectName->Buffer;
+    PCWSTR hklm_path = L"\\Registry\\Machine\\";
+    PCWSTR hkcu_path = L"\\Registry\\User\\";
+    if (relative_path.starts_with(hklm_path)) {
+        root_key = HKEY_LOCAL_MACHINE;
+        relative_path = relative_path.substr(wcslen(hklm_path));
+    } else if (relative_path.starts_with(hkcu_path)) {
+        root_key = HKEY_CURRENT_USER;
+        relative_path = relative_path.substr(wcslen(hkcu_path));
+    }
+
+    ULONG result = RegCreateKeyExW(
+        root_key,
+        relative_path.c_str(),
+        0, // Reserved
+        (class_string ? class_string->Buffer : nullptr),
+        REG_OPTION_VOLATILE,
+        desired_access,
```

Comment:
Could this be unreliable if two different code paths request access to the same key with different desired access masks?

E.g., caller A requests read access and gets HKLM, and caller B requests read/write access and gets HKCU: the expectation is that both callers have a handle to the same key.